accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -291,18 +333,49 @@ private void forEach(Routable keyOrRange, Ranges slice, Consumer<RoutableKey> fo
         }
     }
 
+    @Override
+    protected void registerHistoricalTransactions(Deps deps)
+    {
+        Ranges allRanges = rangesForEpochHolder.get().all();
+        deps.keyDeps.keys().forEach(allRanges, key -> {
+            SafeCommandsForKey cfk = commandsForKey(key).createSafeReference();
+            deps.keyDeps.forEach(key, txnId -> {
+                // TODO (desired, efficiency): this can be made more efficient by batching by epoch
+                if (rangesForEpochHolder.get().coordinates(txnId).contains(key))
+                    return; // already coordinates, no need to replicate
+                if (!rangesForEpochHolder.get().allBefore(txnId.epoch()).contains(key))
+                    return;
+
+                cfk.registerNotWitnessed(txnId);
```

Comment:
it's not clear why we're registering txnIds for ranges we used to replicate.

Reply:
we might not have witnessed it?

Reply:
Sure, but we're not registering txnIds for ranges we didn't used to replicate (asking in the context of bootstrap) so why the distinction there?